### PR TITLE
Add h5py

### PR DESCRIPTION
**Pull request by jakirkham**
Adds a recipe to build h5py. Largely based off of the one in conda-recipes. However, tweaked to work nicely in conda-forge.

### CODE REVIEW

File: recipes/h5py/build.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+export HDF5_DIR="${PREFIX}"
```
Comment (jakirkham) [inline recipes/h5py/build.sh]: Needs to know where to find HDF5.

File: recipes/h5py/build.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+"${PYTHON}" setup.py install
```
Comment (ocefpaf) [inline recipes/h5py/build.sh]: I thin that [h5py](https://github.com/h5py/h5py/blob/master/setup.py#L12) can use setuptools so we can use ` --single-version-externally-managed  --record record.txt`

File: recipes/h5py/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    - six
+    - unittest2    #[py26]
+
+test:
+  imports:
+    - h5py
+
+about:
+  home: http://www.h5py.org/
+  license: BSD 3-Clause
```
Comment (ocefpaf) [inline recipes/h5py/meta.yaml]: :+1:

File: recipes/h5py/build.sh (hunk 2 above)
Comment (jakirkham) [inline recipes/h5py/build.sh]: Yes, you are correct. Will update.

File: recipes/h5py/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    - cython
+    - six
+    - pkgconfig
+
+  run:
+    - python
+    - numpy
+    - hdf5
+    - six
+    - unittest2    #[py26]
```
Comment (jakirkham) [inline recipes/h5py/meta.yaml]: Probably not necessary, but nice.

File: recipes/h5py/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+test:
+  imports:
+    - h5py
```
Comment (jakirkham) [inline recipes/h5py/meta.yaml]: To fool the linter as it won't recognize `run_test.py`.

File: recipes/h5py/run_test.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import h5py.h5o
+import h5py.h5p
+import h5py.h5r
+import h5py.h5s
+import h5py.h5t
+import h5py.h5z
+import h5py.utils
+
+from sys import exit
+exit(0) if h5py.run_tests().wasSuccessful() else exit(1)
```
Comment (jakirkham) [inline recipes/h5py/run_test.py]: Runs the full test suite. Errors if there was a failure. The suite is pretty fast so no point in not running it.

File: recipes/h5py/bld.bat [hunk 7]
```diff
@@ -0,0 +1,4 @@
+SET HDF5_DIR="%PREFIX%"
```
Comment (jakirkham) [inline recipes/h5py/bld.bat]: Think this is needed for Windows too, but was missing from conda-recipes.

File: recipes/h5py/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{% set version = "2.6.0" %}
```
Comment (jakirkham) [inline recipes/h5py/meta.yaml]: This is a newer version that the one in `defaults` or `anaconda`.

File: recipes/h5py/meta.yaml (hunk 4 above)
Comment (ocefpaf) [inline recipes/h5py/meta.yaml]: I am fine with that. Unused in conda-forge, but leaves it general to other users in other places.

File: recipes/h5py/build.sh (hunk 2 above)
Comment (jakirkham) [inline recipes/h5py/build.sh]: Corrected for both.

File: recipes/h5py/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+package:
+  name: h5py
+  version: {{ version }}
+
+source:
+  fn: h5py-{{ version }}.tar.gz
+  url: https://github.com/h5py/h5py/archive/{{ version }}.tar.gz
```
Comment (jakirkham) [inline recipes/h5py/meta.yaml]: Not on PyPI yet, but has been requested. See this issue ( https://github.com/h5py/h5py/issues/669 ) for more info.

File: recipes/h5py/bld.bat (hunk 7 above)
Comment (ocefpaf) [inline recipes/h5py/bld.bat]: You are probably correct. But I guess it won't be %PREFIX%
Maybe  %LIBRARY_PREFIX%?

**Comment by jakirkham** [comment]
Hmm...seeing a single test failure on Linux. Opened this issue ( https://github.com/h5py/h5py/issues/696 ) upstream. Not sure if this isn't due to some weird behavior due to docker.

**Comment by jakirkham** [comment]
The [suggestion]( https://github.com/h5py/h5py/issues/554#issuecomment-95000209 ) by the maintainer is that this test might be broken. Maybe we can mark it for skipping and PR that patch upstream. Though it is concerning that it is not always a problem. Thinking there might be something funny with docker here.

File: recipes/h5py/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  md5: 613c5776192eaf02ff96a82dfc6d2e78
+  patches:
+    # Patch a test that appears to be failing in some cases.
+    # This patch has been submitted to upstream.
+    # See the linked PR for more information.
+    #
+    # https://github.com/h5py/h5py/pull/697
+    #
+    - h5py_tests_old_test_file.py.patch
```
Comment (jakirkham) [inline recipes/h5py/meta.yaml]: Had to add this patch to skip a flaky part of a test. This was submitted upstream in this PR ( https://github.com/h5py/h5py/pull/697 ).

File: recipes/h5py/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/h5py/meta.yaml]: Definitely, I feel the same way. :smile:

File: recipes/h5py/bld.bat [hunk 11]
```diff
@@ -0,0 +1,5 @@
+"%PYTHON%" setup.py configure --hdf5="%LIBRARY_PREFIX%"
```
Comment (jakirkham) [inline recipes/h5py/bld.bat]: Switched to using the [`configure`]( http://docs.h5py.org/en/latest/build.html#custom-installation ) option as setting the environment variable wasn't working on Windows (could have been doing it wrong). This seems a bit nicer and more consistent across platforms.

File: recipes/h5py/build.sh [hunk 12]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+"${PYTHON}" setup.py configure --hdf5="${PREFIX}"
```
Comment (jakirkham) [inline recipes/h5py/build.sh]: Switched to using the [`configure`] option here too.

File: recipes/h5py/bld.bat (hunk 11 above)
Comment (ocefpaf) [inline recipes/h5py/bld.bat]: I like that better too.

PS: I had to set `%LIBRARY_INC%` and `%LIBRARY_LIB%` for `netCDF4-python`.

**Comment by jakirkham** [comment]
Even if this passes on Windows, we should hold off on merging it until we fix this issue ( https://github.com/conda-forge/hdf5-feedstock/issues ) so that we are building against the same (latest) version of HDF5 on all platforms.

**Comment by jakirkham** [comment]
@andrewcollette, we are adding `h5py` to [conda-forge]( https://conda-forge.github.io/ ). This is the community packaging system for use with [`conda`]( http://conda.pydata.org/docs/ ). Packages built here are built with CI and deployed to the [conda-forge channel]( https://anaconda.org/conda-forge/ ) automatically after committing on `master` of a feedstock (a repo with the recipe plus automatically maintained CI scripts). Eventually, packages submitted here will be included in the `defaults` channel after some auditing process as [mentioned]( https://github.com/conda-forge/conda-forge.github.io/issues/22#issuecomment-197117472 ). If there is anything here that you see is off or needs to be fixed, please let me know. If you would like to have access to this feedstock/recipe, I would happily add you to the maintainers list so you can have commit privileges for this recipe. This would be useful for updating the version for instance.

**Comment by jakirkham** [comment]
Windows seems to still be messed up. Given that the HDF5 feedstock is messed up on Windows. ( https://github.com/conda-forge/hdf5-feedstock/issues/1 ), I am inclined to rip out Windows support for now and have someone with a Windows dev environment PR the fix. Would you be ok with this, @ocefpaf?

**Comment by ocefpaf** [comment]
> Windows seems to still be messed up. Given that the HDF5 feedstock is messed up on Windows. ( https://github.com/conda-forge/hdf5-feedstock/issues/1 ), I am inclined to rip out Windows support for now and have someone with a Windows dev environment PR the fix. Would you be ok with this, @ocefpaf?

No need to change the recipe. Leave the Windows stuff there and commit with a `[skip appveyor]` when submitting new changes.

(I like to see the failure to remind us to get back to it. See https://conda-forge.github.io/feedstocks.html)

**Comment by jakirkham** [comment]
> No need to change the recipe. Leave the Windows stuff there and commit with a `[skip appveyor]` when submitting new changes.

The red Xs make me nervous. :fearful:

How about a PR with the changes on the feedstock so we can see them? That way it is a bit more like a task than a broken recipe. Wouldn't want someone to think they could use this to build locally on Windows when they really can't. :wink:

**Comment by ocefpaf** [comment]
> How about a PR with the changes on the feedstock so we can see them? That way it is a bit more like a task than a broken recipe. Wouldn't want someone to think they could use this to build locally on Windows when they really can't. :wink:

:+1:

**Comment by jakirkham** [comment]
Thanks for being understanding.

**Comment by jakirkham** [comment]
@khinsen @FRidh @mwcraig, we are adding `h5py` to [conda-forge]( https://conda-forge.github.io/ ). This is the community packaging system for use with [`conda`]( http://conda.pydata.org/docs/ ). Packages built here are built with CI and deployed to the [conda-forge channel]( https://anaconda.org/conda-forge/ ) automatically after committing on `master` of a feedstock (a repo with the recipe plus automatically maintained CI scripts). Eventually, packages submitted here will be included in the `defaults` channel after some auditing process as [mentioned]( https://github.com/conda-forge/conda-forge.github.io/issues/22#issuecomment-197117472 ). If there is anything here that you see is off or needs to be fixed, please let me know. If you would like to have access to this [feedstock/recipe]( http://github.com/conda-forge/h5py-feedstock/ ), I would happily add you to the maintainers list so you can have commit privileges for this recipe. This would be useful for updating the version for instance.

**Comment by andrewcollette** [comment]
@jakirkham, thanks for pinging me on this.  I will let you guys handle the recipe but am happy to be available if you have questions.

**Comment by jakirkham** [comment]
Sounds good. If you are willing to, you could watch this [feedstock/repo]( https://github.com/conda-forge/h5py-feedstock/ ) to see if there are upstream issues that need to be addressed. Though we will likely raise them and try to give you enough info. Also, if you need to ping about updates to `h5py`, just use this mention @conda-forge/h5py and this will get the attention of anyone in the maintenance team (currently just me).

**Pull request merged**